src/cmd/metadata/mod.rs

```diff
@@ -0,0 +1,367 @@
+        };
+
+        if self.unstable_options.original_manifest {
+            generate_metadata(&self.crate_metadata.manifest_path)?;
+        } else {
+            Workspace::new(&cargo_meta, &self.crate_metadata.root_package.id)?
+                .with_root_package_manifest(|manifest| {
+                    manifest
+                        .with_added_crate_type("rlib")?
+                        .with_profile_release_lto(false)?;
+                    Ok(())
+                })?
+                .with_metadata_gen_package()?
+                .using_temp(generate_metadata)?;
+        }
+
+        Ok(out_path)
+    }
+
+    /// Generate the extended contract project metadata
+    fn extended_metadata(&self) -> Result<(Source, Contract, Option<User>)> {
+        let contract_package = &self.crate_metadata.root_package;
+        let ink_version = &self.crate_metadata.ink_version;
+        let rust_version = Version::parse(&rustc_version::version()?.to_string())?;
+        let contract_name = contract_package.name.clone();
+        let contract_version = Version::parse(&contract_package.version.to_string())?;
+        let contract_authors = contract_package.authors.clone();
+        // optional
+        let description = contract_package.description.clone();
+        let documentation = self.crate_metadata.documentation.clone();
+        let repository = contract_package
+            .repository
+            .as_ref()
+            .map(|repo| Url::parse(&repo))
+            .transpose()?;
+        let homepage = self.crate_metadata.homepage.clone();
+        let license = contract_package.license.clone();
+        let hash = self.wasm_hash()?;
+
+        let source = {
+            let lang = SourceLanguage::new(Language::Ink, ink_version.clone());
+            let compiler = SourceCompiler::new(Compiler::RustC, rust_version);
+            Source::new(hash, lang, compiler)
+        };
+
+        // Required contract fields
+        let contract = Contract::new(
+            contract_name,
+            contract_version,
+            contract_authors,
+            description,
+            documentation,
+            repository,
+            homepage,
+            license,
+        );
+
+        // user defined metadata
+        let user = self.crate_metadata.user.clone().map(User::new);
+
+        Ok((source, contract, user))
+    }
+
+    /// Compile the contract and then hash the resulting wasm
+    fn wasm_hash(&self) -> Result<[u8; 32]> {
+        super::build::execute_with_metadata(
+            &self.crate_metadata,
+            self.verbosity,
+            self.unstable_options.clone(),
+        )?;
+
+        let wasm = fs::read(&self.crate_metadata.dest_wasm)?;
+
+        use ::blake2::digest::{Update as _, VariableOutput as _};
+        let mut output = [0u8; 32];
+        let mut blake2 = blake2::VarBlake2b::new_keyed(&[], 32);
+        blake2.update(wasm);
+        blake2.finalize_variable(|result| output.copy_from_slice(result));
```

Comment:
I think these 3 lines could be replaced with a single call to: https://docs.rs/digest/0.9.0/digest/trait.Digest.html#tymethod.digest